File: Dynamic-program/322-coinChange/322.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    int coinChange(vector<int> &coins, int amount)
    {
        vector<int> dp(amount + 1, INT_MAX);
        dp[0] = 0;
        for (int coin : coins) {
            for (int j = coin; j <= amount; ++j) {
                if (dp[j - coin] < INT_MAX)
                    dp[j] = min(dp[j], dp[j - coin] + 1);
            }
        }
        return dp[amount] == INT_MAX ? -1 : dp[amount];
    }
};

File: Dynamic-program/322-coinChange/go/322.go
package main

import (
	"fmt"
	"sort"
)

func coinChange(coins []int, amount int) int {
	// dp[i]: 凑成 i 需要的最少硬币个数
	sort.Ints(coins)
	dp := make([]int, amount+1)
	for i := range dp {
		dp[i] = amount + 1
	}
	dp[0] = 0
	for i := coins[0]; i <= amount; i++ {
		for j := 0; j < len(coins); j++ {
			if coins[j] <= i {
				dp[i] = min(dp[i], dp[i-coins[j]]+1)
			}
		}
	}
	if dp[amount] > amount {
		return -1
	} else {
		return dp[amount]
	}
}

func main() {
	coins := []int{1, 2, 5}
	fmt.Println(coinChange(coins, 11))
}
